File: Python/SkipList/skiplist.py
## written by xiong biao
## date 2020-5-24

import random

class SkipList(object):

    class Node(object):
        def __init__(self, key, value, level):
            self.key = key
            self.value = value
            self.nexts = [None] * level
            # random.seed(10)

    def __init__(self):
        self._levels = 0
        self._size = 0
        self.MAX_LEVEL = 32
        self.probablity = 0.25
        self.__first = SkipList.Node(None, None, self.MAX_LEVEL)

    def put(self, key, value):
        self.__check_key(key)
        node = self.__first
        prevs = [None] * self._levels
        for i in range(self._levels)[::-1]:
            if node.nexts[i]:
                cmp = node.nexts[i].key - key
            else:
                cmp = -1
            while node.nexts[i] and cmp < 0:
                node = node.nexts[i]
                cmp = node.key - key

            if cmp == 0:
                old_value = node.nexts[i].value
                node.nexts[i].value = value
                return old_value

            prevs[i] = node

        new_level = self.random_level()
        new_node = SkipList.Node(key, value, new_level)

        for i in range(new_level):

            if i >= self._levels:
                self.__first.nexts[i] = new_node
            else:
                new_node.nexts[i] = prevs[i].nexts[i]
                prevs[i].nexts[i] = new_node

        self._size += 1
        self._levels = max(new_level, self._levels)
        return None

    def random_level(self):
        level = 1
        while random.uniform(0, 1) < self.probablity and level < self.MAX_LEVEL:
            level += 1

        return level


    def get(self, key):
        self.__check_key(key)
        node = self.__first
        for i in range(self._levels)[::-1]:
            while node.nexts[i]:
                if node.nexts[i].key == key:
                    return node.nexts[i].value
                if node.nexts[i].key < key:
                    node = node.nexts[i]
        return None

    def remove(self, key):
        self.__check_key(key)
        node = self.__first
        prevs = [None] * self._levels
        exits = False
        for i in range(self._levels)[::-1]:
            if node.nexts[i]:
                cmp = node.nexts[i].key - key
            else:
                cmp = -1
            while node.nexts[i] and cmp < 0:
                node = node.nexts[i]
                cmp = node.key - key

            if cmp == 0:
                exits = True
            prevs[i] = node

        if not exits:
            return None

        remove_node = node.nexts[0]
        for i in range(len(remove_node.nexts)):
            prevs[i].nexts[i] = remove_node.nexts[i]

        new_level = self._levels
        while new_level > 0 and self.__first.nexts[new_level]:
            self._levels = new_level
            new_level -= 1

        self._size -= 1
        return remove_node.value

    def __check_key(self, key):
        if not key:
            raise ValueError('key cannot be None')

    def __repr__(self):
        s = ''

        for i in range(self._levels)[::-1]:
            node = self.__first
            while node.nexts[i]:
                s += str(node.nexts[i].key) + '-' + str(node.nexts[i].value) + '('+str(len(node.nexts[i].nexts))+'), '
                node = node.nexts[i]
            s += '\n'
        return s

sl = SkipList()

for i in range(16):
    sl.put(i, i + 10)

print(sl)
sl.remove(10)
print(sl)

File: Java/src/com/venus/SkipList/SkipList.java
package com.venus.SkipList;

import java.util.Arrays;
import java.util.Comparator;
import java.util.Objects;

@SuppressWarnings("ALL")
public class SkipList<K, V> {
    private static final double probablity = 0.25;
    private static final int maxLevel = 32;
    private int size;
    private int level;
    private Comparator<K> comparator;
    private Node<K, V> first;

    private int compare(K k1, K k2){
        return comparator == null ? ((Comparable<K>)k1).compareTo(k2) : comparator.compare(k1 ,k2);
    }

    public SkipList() {
        this(null);
    }

    public SkipList(Comparator comparator) {
        this.comparator = comparator;
        first = new Node<>(null, null, maxLevel);
    }

    private class Node<K, V> {
        K key;
        V value;
        Node<K, V>[] nexts;
        int index;

        public Node(K key, V value, int level) {
            this.key = key;
            this.value = value;
            nexts = new Node[level];
        }

        @Override
        public String toString() {
            return "key: " + key + ", nexts: " + nexts.length;
        }
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public V get(K key) {
        checkKey(key);
        Node<K, V> node = first;
        for (int i = level - 1 ; i >= 0 ; i--) {
            int cmp = -1;
            while (node.nexts[i] != null && (cmp = compare(node.nexts[i].key, key)) < 0) {
                node = node.nexts[i];
            }
            if (cmp == 0) return node.nexts[i].value;
        }
        return null;
    }

    private int randomLevel() {
        int level = 1;
        while (Math.random() < probablity && level < maxLevel) {
            level ++;
        }
        return level;
    }

    public V put(K key, V value) {
        checkKey(key);
        Node<K, V> node = first;
        Node<K, V>[] prevs = new Node[level];
        for (int i = level - 1 ; i >= 0 ; i--) {
            int cmp = -1;
            while (node.nexts[i] != null && (cmp = compare(node.nexts[i].key, key)) < 0) {
                node = node.nexts[i];
            }
            if (cmp == 0) {
                V oldValue = node.nexts[i].value;
                node.nexts[i].value = value;
                return value;
            }
            prevs[i] = node;
        }

        int newLevel = randomLevel();
        Node<K, V> newNode = new Node<>(key, value, newLevel);

        for (int i = 0; i < newLevel; i++) {
            if (i >= level) {
                first.nexts[i] = newNode;
            }else {
                newNode.nexts[i] = prevs[i].nexts[i];
                prevs[i].nexts[i] = newNode;
            }
        }
        level = Math.max(level, newLevel);
        size ++;
        return null;
    }

    public V remove(K key) {
        checkKey(key);
        Node<K, V> node = first;
        Node<K, V>[] prevs = new Node[level];
        boolean exits = false;
        for (int i = level - 1 ; i >= 0 ; i--) {
            int cmp = -1;
            while (node.nexts[i] != null && (cmp = compare(node.nexts[i].key, key)) < 0) {
                node = node.nexts[i];
            }
            if (cmp == 0) exits = true;
            prevs[i] = node;
        }
        if (!exits) return null;
        Node<K, V> removeNode = node.nexts[0];
        for (int i = 0; i < removeNode.nexts.length; i++) {
            prevs[i].nexts[i] = removeNode.nexts[i];
        }
        int newLevel = level;
        while (--newLevel > 0 && first.nexts[newLevel] != null){
            level = newLevel;
        }
        size -- ;
        return removeNode.value;
    }

    private void checkKey(K key) {
        if (key == null) {
            throw new IllegalArgumentException("key cannot be null..");
        }
    }


    @Override
    public String toString() {
        StringBuilder s = new StringBuilder();
        s.append("Total Level: ").append(level).append("\n");

        for (int i = level - 1; i >= 0; i--) {
            Node<K, V> node = first;
            while (node.nexts[i] != null) {
                s.append(node.nexts[i]);
                s.append("\t");
                node = node.nexts[i];
            }
            s.append("\n");
        }

        return s.toString();
    }
}
